### DISCUSSION TITLE
受项目启发我写了一个用于oagent的CFW脚本，也许可以合并进你的项目。

### DISCUSSION THREAD

**Discussion by s65d4fs654dg** [SeaHOH/GotoX#256] [点子]
其实就是gotox的cfw脚本结合原版oagent的php node.js脚本摘抄整理了下。
放弃了解压缩的方式，经测试解压缩对cpu的占用比较大，很容易超出cfw的限额，把压缩解压缩换成异或加解密。
没使用chunk方式处理数据，直接整个responseBody载入缓存。
要使用修改的客户端，去掉压缩请求头文件的步骤，整个请求数据异或加密。
脚本中也是把请求数据先异或解密，然后根据前两个字节计算请求头文件长度，把请求头文件分离出来。
多数网站没问题，速度也挺快的，不过无法通过google.com的机器人验证(reCAPTCHA)。
```js
'use strict'

var __version__ = '0.8.6-nocompress'
var __password__ = '123456'
var bit = __password__.charCodeAt(0);

addEventListener('fetch', event => event.respondWith(handleRequest(event.request)));

async function handleRequest(request) {

    let originalResponse = await fetch(await readRequest(request));

    let originalResponseClone = originalResponse.clone();
    let responseHeaders = originalResponse.headers;
    let newResponseHeaders = new Headers(responseHeaders);
    let status = originalResponse.status;
		
    newResponseHeaders.delete('Transfer-Encoding');
    newResponseHeaders.delete('Content-Length');
    if(newResponseHeaders.get("x-pjax-url")) { newResponseHeaders.delete("x-pjax-url") };

    var cipher = 1 ;
    let content = 'HTTP/1.1 ' + status + '\r\n';
    for (let [key, value] of newResponseHeaders.entries()) {
        key = key.replace(/\w+/g, function (w) { return w.charAt(0).toUpperCase() + w.substr(1).toLowerCase() });
        if (key == 'Content-Encoding' && value != 'br') { continue };
        if (key == 'Content-Type' && /.*image|video|audio/.test(value.toLowerCase())) { cipher = 0 };
        if (key == 'Set-Cookie') { value = value.toString().replace(/, ([^ =]+(?:=|$))/g, '\r\nSet-Cookie: $1')};
        content += key + ': ' + value + '\r\n';
    }

    let responseBody = new Uint8Array(await originalResponseClone.arrayBuffer());
    let encodedHeaders = new TextEncoder("utf-8").encode(content + '\r\n');
    let res = concatenate(encodedHeaders, responseBody);

    if (cipher == 0) {
        return new Response(res, {
            status: 200,
            statusText: 'OK',
            headers: new Headers({ 'Content-Type': 'image/x-png' })
        })
    } else {
        return new Response(xor(res.buffer, bit), {
            status: 200,
            statusText: 'OK',
            headers: new Headers({ 'Content-Type': 'image/gif' })
        })
	}
}

async function readRequest(encodedRequest) {
    let dataIn = new Uint8Array(await encodedRequest.arrayBuffer());
    let data = new Uint8Array(await (xor(dataIn.buffer, bit)));
    let headers_length = new DataView(data.buffer, 0, 2).getUint16();
    let requestData = new Uint8Array(headers_length);
    requestData.set(data.slice(2, headers_length + 2));
    let requestBody = new Uint8Array(data.length - 2 - headers_length);
    requestBody.set(data.slice(2 + headers_length));
	let requestHeaders =  new Headers();
    let buff = new TextDecoder().decode(requestData);
    var lines = buff.toString().split("\r\n");
    var request_line_items = lines.shift().split(" ");
    var requestMethod = request_line_items[0];
    var requestUrl = request_line_items[1];
    for (var i = 0; i < lines.length; i++) {
        var line = lines[i];
        var pos = line.indexOf(':');
        if (pos > 0) {
            var key = line.substring(0, pos);
            var value = line.substring(pos + 1).trim();
            if (key == "Content-Length") {
                requestHeaders.append('Content-Length',requestBody.length);
            } else {
                requestHeaders.append(key,value);
            }
        }
    }

    let url = new URL(requestUrl)

    if (requestBody.length == 0) {
        const newRequestInit = {
            method: requestMethod,
            headers: requestHeaders,
        }
        return new Request(url, newRequestInit)
    } else {
        const newRequestInit = {
            method: requestMethod,
            headers: requestHeaders,
            body: requestBody,
        }
        return new Request(url, newRequestInit)
    }

}

/*************************************************************************************/

if (!Uint8Array.prototype.slice) {
    Uint8Array.prototype.slice = function (a, b) {
        var Uint8ArraySlice = new Uint8Array(this.buffer.slice(a, b));
        return Uint8ArraySlice;
    }
}

function concatenate(...arrays) {
    let size = arrays.reduce((a, b) => a + b.byteLength, 0);
    let result = new Uint8Array(size);
    let offset = 0;
    for (let arr of arrays) {
        result.set(arr, offset);
        offset += arr.byteLength;
    }
    return result;
}

function xor(buffer, value) {
    const view = new Uint8Array(buffer)
    view.forEach((byte, offset) => view.set([byte ^ value], offset))
    return buffer
}


```
正常长时间使用无cpu内存超限额。
![1](https://user-images.githubusercontent.com/13595037/205547363-71c3e41d-5247-4409-ae95-8b4c7e745bba.PNG)
![2](https://user-images.githubusercontent.com/13595037/205547366-81a59352-fc9b-4bd9-8ab4-6239014bb028.PNG)
![3](https://user-images.githubusercontent.com/13595037/205547371-904c6c20-7fd1-4d32-a3d6-98f0e74419c3.jpg)

**Comment by SeaHOH** [comment]
首先谢谢你的建议，但是我暂时会保持原样。

各人的使用方式上，大概差别还是挺大的，我至今并未触及过 CPU 时间限额。如果有出现超限，我估计一般会是请求数。
之所以采取压缩传输，主要还是为了连接能尽可能快，但因为没啥必要，我并没实际测试过传输延迟。
对比你我的数据，解压缩时平均 Wokers 多耗了2-3ms，如果不触及限额，我认为这是可以接受的。
我没有直接读取 body 是为了尽量兼容以后可能的标准，实际代理请求前也是会完全读取到内存的，除了不能上传大文件其它都还行。

另外给你个建议：因为已经是加密连接，异或操作没有必要，可以取消。

![cfw_cpu_time](https://user-images.githubusercontent.com/8470820/205671173-ee92d622-279d-4a9a-80ff-c79fc31483ae.png)
![cfw_errors](https://user-images.githubusercontent.com/8470820/205671213-efef269c-0089-4fc9-af9e-d5430db8b0a2.png)

**Reply by s65d4fs654dg** [reply]
> 首先谢谢你的建议，但是我暂时会保持原样。
> 
> 各人的使用方式上，大概差别还是挺大的，我至今并未触及过 CPU 时间限额。如果有出现超限，我估计一般会是请求数。 之所以采取压缩传输，主要还是为了连接能尽可能快，但因为没啥必要，我并没实际测试过传输延迟。 对比你我的数据，解压缩时平均 Wokers 多耗了2-3ms，如果不触及限额，我认为这是可以接受的。 我没有直接读取 body 是为了尽量兼容以后可能的标准，实际代理请求前也是会完全读取到内存的，除了不能上传大文件其它都还行。
> 
> 另外给你个建议：因为已经是加密连接，异或操作没有必要，可以取消。
> 
> ![cfw_cpu_time](https://user-images.githubusercontent.com/8470820/205671173-ee92d622-279d-4a9a-80ff-c79fc31483ae.png) ![cfw_errors](https://user-images.githubusercontent.com/8470820/205671213-efef269c-0089-4fc9-af9e-d5430db8b0a2.png)

的确，实际情况不一样，我是因为cpu经常超限才修改的。
另外有没有考虑加入http直连支持，cfw用http(非https)是可以直连的。异或加密也是因为我用http非加密才加上去的，包括响应数据的加密。

**Reply by SeaHOH** [reply]
> 另外有没有考虑加入http直连支持，cfw用http(非https)是可以直连的。异或加密也是因为我用http非加密才加上去的，包括响应数据的加密。

这确实是个盲点，谢谢提醒！我会抽空进行测试，看能否安全的利用上。

**Comment by SeaHOH** [comment]
>> 另外有没有考虑加入http直连支持，cfw用http(非https)是可以直连的。异或加密也是因为我用http非加密才加上去的，包括响应数据的加密。
>
>这确实是个盲点，谢谢提醒！我会抽空进行测试，看能否安全的利用上。

如果使用 http 连接附带自定义的加密方案，因为 Workers 机制不支持流式处理返回结果 (直接返回原始 `Response` 和 HTML 专用的 `HTMLRewriter` 是支持流式的)，这里有两个问题：
1. WebSocket 肯定是废了，它需要保持连接；
1. 响应结果越大延迟越大，且极不确定，受多方因素影响。

基于此，我不会添加 Workers 的 http 连接方式。
当然，个人这样使用大概是没有问题的，因为可以不加密 WS 连接和大文件，你贴图中的视频数据大概也是未混淆的吧。

**Reply by SeaHOH** [reply]
http 有搞头！突然发现，从去年开始[可以玩 WebSocket 了](https://developers.cloudflare.com/workers/runtime-apis/websockets/)。

**Comment by SeaHOH** [comment]
当前采用的 inflate 来自 [imaya](https://github.com/imaya/zlib.js)，这几天我对比测试了 [pako](https://github.com/nodeca/pako) ，在当前使用条件下 (绝大多数数据原始大小位于 100-3000 区间)，大概是这样：
- imaya 的时间曲线相比 pako 更为陡峭
- 当数据大小在 1000 以下时，imaya 耗时更少
- 当数据大小在 2000 以上时，pako 耗时更少
- 当数据大小在 3000 以内 (大约两个 MUT)，是否压缩数据对延迟影响不大，不压缩有时反而会稍快些 (10ms 以内)
- 更大的数据未测试，因为暂时没有发现合适的站点，也懒得专门写个测试用的 Worker

基于以上结果，接下来会进行以下修改：
- 替换并使用 pako
- 3000 以内的数据不再压缩
